mlflow/store/artifact/databricks_artifact_repo.py

```diff
@@ -207,26 +256,29 @@ def _azure_upload_file(self, credentials, local_file, artifact_path):
         """
         try:
             headers = self._extract_headers_from_credentials(credentials.headers)
-            uploading_block_list = []
-            for chunk in yield_file_in_chunks(local_file, _AZURE_MAX_BLOCK_CHUNK_SIZE):
-                # Base64-encode a UUID, producing a UTF8-encoded bytestring. Then, decode
-                # the bytestring for compliance with Azure Blob Storage API requests
-                block_id = base64.b64encode(uuid.uuid4().hex.encode()).decode("utf-8")
-                try:
-                    put_block(credentials.signed_uri, block_id, chunk, headers=headers)
-                except requests.HTTPError as e:
-                    if e.response.status_code in [401, 403]:
-                        _logger.info(
-                            "Failed to authorize request, possibly due to credential expiration."
-                            " Refreshing credentials and trying again..."
-                        )
-                        credential_info = self._get_write_credential_infos(
-                            run_id=self.run_id, paths=[artifact_path]
-                        )[0]
-                        put_block(credential_info.signed_uri, block_id, chunk, headers=headers)
-                    else:
-                        raise e
-                uploading_block_list.append(block_id)
+            futures = {}
+            num_chunks = _compute_num_chunks(local_file, _MULTIPART_UPLOAD_CHUNK_SIZE)
+            for index in range(num_chunks):
+                start_byte = index * _MULTIPART_UPLOAD_CHUNK_SIZE
+                future = self.thread_pool.submit(
```

Comment:
boto3 limits the number of chunks that can be stored in memory:

https://github.com/boto/s3transfer/blob/a3a851f1df5dd5f59b319d626a44f4c8893d7113/s3transfer/manager.py#L105

```
        :param max_in_memory_upload_chunks: The number of chunks that can
            be stored in memory at a time for all ongoing upload requests.
            This pertains to chunks of data that need to be stored in memory
            during an upload if the data is sourced from a file-like object.
            The total maximum memory footprint due to a in-memory upload
            chunks is roughly equal to:
```

Reply:
boto3 reads chunks before submitting jobs, so the task queue must be bounded.